zerver/lib/user_groups.py

```diff
@@ -46,17 +68,57 @@ def access_user_group_by_id(
     return user_group
 
 
-def access_user_groups_as_potential_subgroups(
-    user_group_ids: Sequence[int], acting_user: UserProfile
-) -> List[UserGroup]:
-    user_groups = UserGroup.objects.filter(id__in=user_group_ids, realm=acting_user.realm)
+@contextmanager
+def lock_subgroups_with_respect_to_supergroup(
+    potential_subgroup_ids: Collection[int], potential_supergroup_id: int, acting_user: UserProfile
+) -> Iterator[LockedUserGroupContext]:
+    """This locks the user groups with the given potential_subgroup_ids, as well as their indirect subgroups,
+    followed by the potential supergroup. It ensures that we lock the user groups in a consistent order
+    topologically to avoid unnecessary deadlocks on non-conflicting queries.
```

Comment:
These lines are too long; we usually wrap comments/docstrings to 80 characters, not 100.